keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1108 @@
+
+  Example pod spec:
+
+  ```yaml
+  apiVersion: v1
+  kind: Pod
+  spec:
+    resources:
+      requests:
+        memory: 100Mi
+        cpu: 2
+      limits:
+        memory: 200Mi
+        cpu: 3
+    initContainers:
+    - name: init-myservice
+      resources:
+        requests:
+          memory: 300Mi
+          cpu: 4
+        limits:
+          memory: 600Mi
+          cpu: 6
+    containers:
+    - name: container-1
+      resources:
+        requests:
+          memory: 50Mi
+          cpu: 1
+        limits:
+          memory: 100Mi
+          cpu: 2
+      - name: container-2
+        resources:
+          requests:
+            memory: 50Mi
+            cpu: 1
+          limits:
+            memory: 100Mi
+            cpu: 2 
+  ```
+
+  **Resources requests at pod-level**
+
+  * The scheduler will use the higher of init and pod-level requests values to find the suitable node for pod placement.
+    
+    From the example above, the scheduler will use:
+
+      max(init container resources requests, pod-level resources requests) 
+
+        = max (memory: 300Mi cpu: 4, memory: 100 Mi cpu: 2)
+
+        = (memory: 300 Mi, cpu: 4) to find a suitable node.
+
+  * The init containers will rely on its own requests specified in the PodSpec. 
+  * After the init containers execution is completed, kubelet will use the pod-level CPU request to set cpu.weight at pod-level cgroup.
+
+  **Resources limits at pod-level**
+
+  * The init containers will rely on their own limits specified in the PodSpec.
+  * After the init containers execution is completed, kubelet will use CPU and memory limits at pod-level, and translate them to cpu.max and memory.max cgroup settings
+  at pod level respectively. 
+
+  **Pros**
+
+  * Reduces the problem of overprovisioning.
+
+  **Cons**
+
+  * Complicates the definitions of QoS classes as this will require considering requests/limits for both initContainer and pod-level requests/limits to attach a 
+  QoS class with a pod. It might be worth setting dynamic QoS classes if this option were to be implemented. The QoS class in the initialization phase is based on
+  init containers requests/limits, while the QoS class after the initialization phase takes pod-level requests/limits into account. 
+  * Complicated to implement as it requires additional logic in kubelet to monitor init containers completion after which it will set/change pod level cgroup settings.
+
+**Note: Option 1 seems easier to implement in the alpha phase to gather feedback from the users. If there are enough supporting use cases for Option 2, we can explore 
+the implementation in the beta phase of the KEP. Option 2 will also require discussions about dynamic QoS classes as per the execution phase of the pod 
+(init vs non-init phase).**
+
+
+#### Admission Controller
+
+* LimitRanger
```

Comment:
do you want to have a defaulter for Pod resources as well? Or this will be out of scope?

Reply:
Good question. 

Today, in the LimitRanger, you specify the "container" level resource requirements. https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/#create-a-limitrange-and-a-pod
We shouldn't change how that works since it has nothing to do with the pod level resource.

What we can do is to extend the LimitRanger to allow `type: Pod` as well for users who'd like to apply pod level defaults.

Reply:
+1 on what Yuju suggested above.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L5673 LimitRanger already supports Pod type today. Reframed this section to add more clarity.

Reply:
will we implement any logic in limit ranger to account for the pod level requests? If Pod has minimum set, should limit ranger still set minimums on container level? Or it should refrain from updating container-level values?

What if there is a defaulter installed and somebody trying to us this feature. They will be forced to specify some small requests on containers to avoid defaulter overriding those values. Is it OK to force people to set those small values? If so - we need to document it into the rollout section

Reply:
Since limit ranger is implemented as a list, I think it's better to keep different limitRangers independent. 

For alpha, we can suggest users to use a different namespace if they don't want the default limitRanger for containers.

Reply:
If there's a limit ranger of type container, it will continue to update the container-level resources even if the user uses pod-level resources. If the user doesn't want to use the container-level defaulting, they should create pods in a different namespace.